### DISCUSSION TITLE
client.publish error message "TypeError: object of type 'float' has no len()"

### DISCUSSION THREAD

**Discussion by Johnl52** [micropython/micropython#10919] [RP2040 / Pico]
I have a pico w with a BME280, a CMPS12 compass and a OLED display, I can read the sensors and display the values on the OLED.

I'm now trying to publish the values to a MQTT server. When I run the code I get the error message 

connected
ip = 192.168.1.163
Temperature:  69.48   Humidity:  27.56   Pressure:  29.94
heading= 207.2
swarm/temp
69.476
Traceback (most recent call last):
  File "<stdin>", line 108, in <module>
  File "<stdin>", line 56, in publish
  File "/lib/umqtt/simple.py", line 112, in publish
TypeError: object of type 'float' has no len()

Here is the code

```
#added mqtt to code reading BME280, CMPS12 compass
#and display on OLED
#sending data to the mosquitto server

from machine import Pin, I2C
from ssd1306 import SSD1306_I2C
from time import sleep
import bme280
import network
import time
from umqtt.simple import MQTTClient

ssid = "Fios-tCL9H_EXT"
password = "PamPolly_1155"

mqtt_server = "192.168.1.171"

#Our wlan_connect function
def wlan_connect():
    wlan = network.WLAN(network.STA_IF)
    wlan.active(True)
    wlan.connect(ssid, password)

    max_wait = 10
    while max_wait > 0:
        if wlan.status() < 0 or wlan.status() >= 3:
            break
        max_wait -= 1
        print('waiting for connection...')
        time.sleep(1)
        
    if wlan.status() != 3:
        raise RuntimeError('network connection failed')
    else:
        print('connected')
        status = wlan.ifconfig()
        print( 'ip = ' + status[0] )
        
def callback(topic, msg):
    print("message received")
    print(msg.decode('utf-8'))
    

def mqtt_connect():
    #mqtt client
    client = MQTTClient(client_id='swarm_picow', server=mqtt_server, port=1883, user=None, password=None,keepalive=3600,ssl=False, ssl_params={})
    client.set_callback(callback)
    client.connect()
    return(client)

client = mqtt_connect()

def publish(topic, value):
    print(topic)
    print(value)
    client.publish(topic, value)
    print("Publish Done")
    
i2c=I2C(0,sda=Pin(0), scl=Pin(1), freq=400000)
oled = SSD1306_I2C(128, 64, i2c)
bme = (bme280.BME280(i2c=i2c))

def read_bytes(i2cAddr, regAddr, buf):
    cmd    = bytearray(1)
    cmd[0] = regAddr & 0xff
    i2c.writeto(i2cAddr, cmd, False)
    i2c.readfrom_into(i2cAddr, buf)
    

def get_heading(hires):
    """ Returns heading with or w/o tilt compensation and/or calibration,
        if available.
        NOTE: The CMPS12 has built-in tilt compensation and is pre-calibra-
        ted, therefore the parameters "tilt" and "calib" are only for
        compatibility reasons and have no effect. With "hires"=True, the
        precision is 3599/360°, otherwise 255/360°.
    """
    ADDRESS_CMPS12 = const(0x60)
    REG_BEARING_8BIT = const(0x01)
    REG_BEARING_16BIT_HB = const(0x2)
    
    hd = 1
    if hires:
      buf = bytearray(2)
      read_bytes(ADDRESS_CMPS12, REG_BEARING_16BIT_HB, buf)
      #print("buf=",buf)
      hd  = ((buf[0] << 8) | buf[1]) /10
    else:
      buf = bytearray(1)
      read_bytes(ADDRESS_CMPS12, REG_BEARING_8BIT, buf)
      hd  = buf[0]/255 *360
    return hd

# connect to Wifi
wlan_connect()

while True:
    temp, pressure, humidity = bme.read_compensated_data()
    temp_f = ((temp * 1.8) / 100) + 32
    press_Hg = (pressure/256) / 3386.39
    humd = humidity/1024
    reading = 'Temperature: ' + f'{(temp_f): .2f}' + '   Humidity: ' + f'{(humd): .2f}' + '   Pressure: ' + f'{(press_Hg): .2f}'
    print(reading)
    head = get_heading(True)
    print ("heading=",head)
    
    #send to mqtt server
    publish('swarm/temp', temp_f)
    Spublish('swarm/humd', humd)
    publish('swarm/press', press_Hg)
    publish('swarm/head', head)
    
    oled.fill(0)
    # Show values (data, x, y)
    oled.text("Temp: ", 1, 5 )
    oled.text(f'{(temp_f): .2f}', 50, 5)
    oled.text("Press: ", 1, 20 )
    oled.text(f'{(press_Hg): .2f}', 50, 20)
    oled.text("Humid: ", 1, 35 )
    oled.text(f'{(humd): .2f}', 50, 35)
    oled.text("Head: ", 1, 50 )
    oled.text(f'{(head): .2f}', 50, 50)
    oled.show()
    sleep(1)

```

Why do I get that error?

Thanks 
John

**Comment by bixb922** [comment]
temp_f, press_Hg and humd are floating point variables. Mqtt publish wants a bytearray, so you should pass ```str(temp_f).encode()``` to mqtt publish. Same with the other floating point numbers.

**Comment by sosi-deadeye** [comment]
Drop-in replacement:

```python
def publish(topic, value):
    if not isinstance(value, bytes):
        value = str(value).encode()
        
    print(f"{topic} -> {value}")
    client.publish(topic, value)
    print("Publish Done")
```

It's also possible to use struct to send int, float, etc.

```python
def struct_publish(topic, value):
    if isinstance(value, float):
        data = struct.pack("<f", value) # float, 4 bytes
    elif isinstance(value, int):
        data = struct.pack("<i", value) # integer, 4 bytes
    else:
        data = str(value).encode("ascii")
        
    print(f"{topic} -> {value}")
    client.publish(topic, value)
    print("Publish Done")
```

**Comment by Johnl52** [comment]
Excellent, that did it, both the way @bixb922 explained and the first example from @sosi-deadeye. I knew it had to do with the variables being floats so I had tried making them integers, but it didn't like that either, I didn't think about making it a string.

One last question, I want to format the value I'm sending to mqtt as a float with 2 decimal places. I tried using the format I have in displaying the full reading,

reading = 'Temperature: ' + f'{(temp_f): .2f}'

but I've tried it several ways, mainly because I'm not sure of the proper way to configure that, and it's not working. Would I use this method or is there a different way?

Thanks for the help and the education.
John

**Comment by bixb922** [comment]
What problem is there when you say it is not working?

F-strings are a good way to format, you can shorten to: ```reading = f'Temperature: {temp_f: .2f}'```, but what you wrote seems correct. 

If you need to send more complex data, there is a good review of serialization here (3.2): https://github.com/peterhinch/micropython-samples

**Comment by Johnl52** [comment]
Sorry I should have been clearer, using the format in the reading line of "f'{(temp_f): .2f}'" I replaced the temp_f with the string conversion to this "publish('swarm/temp', f'{(str(temp_f).encode()): .2f}')" and I get the following error

  File "<stdin>", line 117, in <module>
ValueError: unknown format code 'f' for object of type 'bytes'

obviously that is not the correct way. I tried the same format and set it to a second variable to publish but I get the same error.

Thanks

**Comment by bixb922** [comment]
It's either ```str(temp_f).encode()``` or ```f'{temp_f: .2f}'.encode()```

Both str and the F-string convert to string. Once you have the string, you encode it.

```f'{(str(temp_f).encode()): .2f}'``` would mean "convert a floating point variable to string, then encode that as a bytearray and then try to format the bytearray as if it were a floating point variable".

**Comment by Johnl52** [comment]
Perfect, I had tried it with just the temp_f previously but I think I had the .encode() and the.2F reversed.

It's hard trying to figure this stuff out on your own when all you have is some Google searches and YouTube videos. Although I've learnt a lot that way, I always get great help from the forums.

Thanks again, greatly appreciated.
John
